src/cascadia/TerminalConnection/AzureConnection.cpp

```diff
@@ -0,0 +1,478 @@
+        responseT.wait();
+        http_response response = responseT.get();
+        pplx::task<json::value> responseJsonT = response.extract_json();
+        responseJsonT.wait();
+        return responseJsonT.get();
+    }
+
+    // Helper function to start the device code flow
+    json::value AzureConnection::_getDeviceCode()
+    {
+        //Initialize the client
+        http_client loginClient(_loginUri);
+
+        //Initialize the request
+        http_request commonRequest(L"POST");
+        commonRequest.set_request_uri(L"common/oauth2/devicecode");
+        utility::string_t body = L"client_id=" + AzureClientID + L"&resource=" + _wantedResource;
+        commonRequest.set_body(body, L"application/x-www-form-urlencoded");
+
+        //Send the request and receive the response as a json value
+        return _requestHelper(loginClient, commonRequest);
+    }
+
+    // Helper function to wait for the user to authenticate using their web browser
+    json::value AzureConnection::_waitForUser(const utility::string_t deviceCode, int pollInterval, int expiresIn)
+    {
+        //Initialize the client
+        http_client pollingClient(_loginUri);
+
+        //Continuously send a poll request until the user authenticates
+        utility::string_t body = L"grant_type=device_code&resource=" + _wantedResource + L"&client_id=" + AzureClientID + L"&code=" + deviceCode;
+        utility::string_t requestUri = L"common/oauth2/token";
+        json::value responseJson;
+        for (int count = 0; count < expiresIn / pollInterval; count++)
+        {
+            // User might close the tab while we wait for them to authenticate, this case handles that
+            if (_closing.load())
+            {
+                throw "Tab closed.";
+            }
+            http_request pollRequest(L"POST");
+            pollRequest.set_request_uri(requestUri);
+            pollRequest.set_body(body, L"application/x-www-form-urlencoded");
+
+            responseJson = _requestHelper(pollingClient, pollRequest);
+
+            if (responseJson.has_field(L"error"))
+            {
+                Sleep(pollInterval * 1000); //Sleep takes arguments in milliseconds
+                continue; //Still waiting for authentication
+            }
+            else
+            {
+                _outputHandlers(winrt::to_hstring("Authenticated.\r\n"));
+                break; //Authentication is done, break from loop
+            }
+        }
+        if (responseJson.has_field(L"error"))
+        {
+            throw "Time out.";
+        }
+        return responseJson;
+    }
+
+    // Helper function to acquire the user's Azure tenants
+    json::value AzureConnection::_getTenants()
+    {
+        //Initialize the client
+        http_client tenantClient(_resourceUri);
+
+        //Initialize the request
+        http_request tenantRequest(L"GET");
+        tenantRequest.set_request_uri(L"tenants?api-version=2018-01-01");

```

Comment:
I swear the other ones above had an API version of 2019-01-01. Is this an error?

Reply:
There's more below here that are 2018-01-01 too. Might be fine, it just stuck out to me.

Reply:
Perhaps `"api-version=2018-01-01"` should be a constant somewhere

Reply:
When we are sending a request to _resourceUri (https://management.azure.com/) we use "api-version=2018-01-01". When we send a request to _cloudShellUri (which is a uri specific to our session), we use "version=2019-01-01" (note that it does not have "api" in front of it).

Reply:
So it's intentional that they're different? It just looks like an accident/typo.